Untangle pkg/client/cli

LukeShu opened the pull request:
## Description

This is a cleanup of `pkg/client/cli/`.  The big ticket item is moving everything over to `cliutil`, so there aren't 2 different `WithConnector` functions.

The total diff is pretty big, but if you step through it commit-by-commit, it should be manageable.  Each commit is pretty small.

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [x] I made sure to update `./CHANGELOG.md`. - yes
 - [ ] I made sure to either submit a docs PR, or tell Matt about the necessary documentation changes. - only docs thing is the release notes
 - [x] My change is adequately tested. - CI should cover it
 - [x] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently. - no tricks

LukeShu committed:
telepresence_test.go: Always run telepresence out-of-process

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
cli: Rewrite daemon_state.go for cliutil, have everything use it

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
cli: current-cluster-id: Simplify implementation, improve error reporting

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
cli: Fix races with testing if the connector is running

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
Delete inaccurate comment

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
cli: Get rid of connectorState and sessionInfo in favor of cliutil

This patch looks big, but most of it is just search/replace noise.

Interesting parts are:
 - cliutil/connector.go
 - cliutil/manager.go
 - session.go

A few of the files benefit from 'ignore whitespace change' when
viewing the diff:
 - preview.go
 - session.go
 - intercept_state.go

The interesting stuff is in `session.go`.

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
Nothing ever sets DaemonStatus.Error; get rid of it

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
Use dlib/dcontext.WithoutCancel

https://github.com/datawire/dlib/pull/21

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
.golangci.yml: Sort the list of linters

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
.golangci.yml: Turn on nolintlint

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
.golangci.yml: Remove unsed exception

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
cli: intercept_state.go: Clean up arg parsing

- Rename 'interceptInfo' to 'interceptArgs' and have it be *just* a
  plain struct of the parsed flags and arguments.

- Create a separate anonymouse RunE wrapper to handle the arg parsing,
  separate the parsing from the code of actually creating the
  intercept.

- Move loginIfNeeded out to be a proper top-level function, rather
  than a closure.

All in all, this is pretty mundane and is a lot of noise with adding
'.args' or '.cs' in to a bunch of accesses on interceptState.

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
Clean up connector intercept error codes

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
cli: intercept_state.go: Prepare for the removal of connectorState

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
cli: session.go: Improve comment

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
pkg/client/cli: Move filenames around, move things between files

Absolutely no code is changing in this commit.  I'm just organizing
things.

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
logging: Adjust the log formatter to work better with dexec

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

thallgren reviewed (approved):
Well needed cleanup for sure. All of it looks good to me.

thallgren commented on pkg/client/cli/intercept_state.go:
```diff
@@ -29,33 +29,34 @@ import (
+	localOnly   bool   // --local-only
+
+	previewEnabled bool                 // --preview-url // only valid if !localOnly
+	previewSpec    *manager.PreviewSpec // --preview-url-* // only valid if !localOnly
```
This should probably not be a pointer given how it's used (causes the test failure).

LukeShu replied:
Yeah, I had to adjust it to get initialized... the linter was unhappy with it not being a pointer because I guess it contains a lock, so it didn't like being copied by value.

thallgren reviewed (changes requested):
The build found an error that I missed.

LukeShu committed:
telepresence_test.go: Avoid hang when deleting namespaces

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu commented:
Changes from the version that @thallgren initially reviewed:
```patch
diff --git a/cmd/telepresence/main.go b/cmd/telepresence/main.go
index 68525f52..5b69e9e3 100644
--- a/cmd/telepresence/main.go
+++ b/cmd/telepresence/main.go
@@ -6,10 +6,18 @@ import (
 	"os"
 
 	"github.com/telepresenceio/telepresence/v2/pkg/client/cli"
+	"github.com/telepresenceio/telepresence/v2/pkg/filelocation"
 )
 
 func main() {
 	ctx := context.Background()
+	if dir := os.Getenv("DEV_TELEPRESENCE_CONFIG_DIR"); dir != "" {
+		ctx = filelocation.WithAppUserConfigDir(ctx, dir)
+	}
+	if dir := os.Getenv("DEV_TELEPRESENCE_LOG_DIR"); dir != "" {
+		ctx = filelocation.WithAppUserLogDir(ctx, dir)
+	}
+
 	cmd := cli.Command(ctx)
 	if err := cmd.ExecuteContext(ctx); err != nil {
 		fmt.Fprintf(cmd.ErrOrStderr(), "%s: error: %v\n", cmd.CommandPath(), err)
diff --git a/pkg/client/cli/cliutil/daemon.go b/pkg/client/cli/cliutil/daemon.go
index 6d2f688e..c130a005 100644
--- a/pkg/client/cli/cliutil/daemon.go
+++ b/pkg/client/cli/cliutil/daemon.go
@@ -17,6 +17,7 @@ import (
 	"google.golang.org/grpc"
 	empty "google.golang.org/protobuf/types/known/emptypb"
 
+	"github.com/datawire/dlib/dexec"
 	"github.com/telepresenceio/telepresence/rpc/v2/daemon"
 	"github.com/telepresenceio/telepresence/v2/pkg/client"
 	"github.com/telepresenceio/telepresence/v2/pkg/filelocation"
@@ -25,6 +26,8 @@ import (
 var ErrNoDaemon = errors.New("telepresence root daemon is not running")
 
 func launchDaemon(ctx context.Context, dnsIP string) error {
+	fmt.Println("Launching Telepresence Daemon", client.DisplayVersion())
+
 	// Ensure that the logfile is present before the daemon starts so that it isn't created with
 	// root permissions.
 	logDir, err := filelocation.AppUserLogDir(ctx)
@@ -53,6 +56,26 @@ func launchDaemon(ctx context.Context, dnsIP string) error {
 
 	args := []string{client.GetExe(), "daemon-foreground", logDir, configDir, dnsIP}
 	if os.Geteuid() != 0 {
+		// If we're going to be prompting for the `sudo` password, we want to first provide
+		// the user with some info about exactly what we're prompting for.  We don't want to
+		// use `sudo`'s `--prompt` flag for this because (1) we don't want it to be
+		// re-displayed if they typo their password, and (2) it might be ignored anyway
+		// depending on `passprompt_override` in `/etc/sudoers`.  So we'll do a pre-flight
+		// `sudo --non-interactive true` to decide whether to display it.
+		needPwCmd := dexec.CommandContext(ctx, "sudo", "--non-interactive", "--true")
+		needPwCmd.DisableLogging = true
+		if err := needPwCmd.Run(); err != nil {
+			fmt.Printf("Need root privileges to run: %s\n", shellquote.Join(args...))
+			// `sudo` won't be able to read the password from the terminal when we run
+			// it with Setpgid=true, so do a pre-flight `sudo true` to read the
+			// password, and then enforce that being re-used by passing
+			// `--non-interactive`.
+			pwCmd := dexec.CommandContext(ctx, "sudo", "true")
+			pwCmd.DisableLogging = true
+			if err := pwCmd.Run(); err != nil {
+				return err
+			}
+		}
 		args = append([]string{"sudo", "--non-interactive", "--preserve-env"}, args...)
 	}
 
@@ -145,6 +168,7 @@ func DidLaunchDaemon(ctx context.Context) bool {
 
 func QuitDaemon(ctx context.Context) error {
 	err := WithStartedDaemon(ctx, func(ctx context.Context, daemonClient daemon.DaemonClient) error {
+		fmt.Print("Telepresence Daemon quitting...")
 		_, err := daemonClient.Quit(ctx, &empty.Empty{})
 		return err
 	})
@@ -157,5 +181,6 @@ func QuitDaemon(ctx context.Context) error {
 		}
 		return err
 	}
+	fmt.Println("done")
 	return nil
 }
diff --git a/pkg/client/cli/cmds_intercept.go b/pkg/client/cli/cmds_intercept.go
index 410d95de..16e6eef5 100644
--- a/pkg/client/cli/cmds_intercept.go
+++ b/pkg/client/cli/cmds_intercept.go
@@ -120,6 +120,7 @@ func interceptCommand(ctx context.Context) *cobra.Command {
 		`Generate an edgestack.me preview domain for this intercept. `+
 		`(default "true" if you are logged in with 'telepresence login', default "false" otherwise)`,
 	)
+	args.previewSpec = &manager.PreviewSpec{}
 	addPreviewFlags("preview-url-", flags, args.previewSpec)
 
 	flags.StringVarP(&args.envFile, "env-file", "e", "", ``+
diff --git a/pkg/client/cli/telepresence_test.go b/pkg/client/cli/telepresence_test.go
index 3d93d779..bfbf6317 100644
--- a/pkg/client/cli/telepresence_test.go
+++ b/pkg/client/cli/telepresence_test.go
@@ -5,7 +5,6 @@ import (
 	"context"
 	"encoding/base64"
 	"fmt"
-	"io"
 	"io/ioutil"
 	"log"
 	"net"
@@ -31,7 +30,7 @@ import (
 	"github.com/datawire/dlib/dhttp"
 	"github.com/datawire/dlib/dlog"
 	"github.com/telepresenceio/telepresence/v2/pkg/client"
-	"github.com/telepresenceio/telepresence/v2/pkg/client/cli"
+	_ "github.com/telepresenceio/telepresence/v2/pkg/client/cli"
 	"github.com/telepresenceio/telepresence/v2/pkg/filelocation"
 	"github.com/telepresenceio/telepresence/v2/pkg/version"
 )
@@ -160,6 +159,7 @@ func (ts *telepresenceSuite) TearDownSuite() {
 	ctx := dlog.NewTestContext(ts.T(), false)
 	_ = run(ctx, "kubectl", "config", "use-context", "default")
 	_ = run(ctx, "kubectl", "delete", "namespace", ts.namespace)
+	_ = run(ctx, "kubectl", "delete", "mutatingwebhookconfiguration", "agent-injector-webhook-"+ts.managerTestNamespace)
 	_ = run(ctx, "kubectl", "delete", "namespace", ts.managerTestNamespace)
 	// Undo RBAC things
 	_ = run(ctx, "kubectl", "delete", "-f", "k8s/client_rbac.yaml")
@@ -546,39 +546,42 @@ func (cs *connectedSuite) TestK_DockerRun() {
 	require.NoError(err)
 	abs, err := filepath.Abs(testDir)
 	require.NoError(err)
-	// Kill container on exit
-	defer func() {
-		_ = dexec.CommandContext(ctx, "docker", "kill", fmt.Sprintf("intercept-%s-%s-8000", svc, cs.ns())).Run()
-	}()
 
-	ccx, cancel := context.WithCancel(ctx)
-	stdoutCh := make(chan string)
-	go func() {
-		stdout, _ := telepresenceContext(ccx, "intercept", "--namespace", cs.ns(), svc,
+	grp := dgroup.NewGroup(ctx, dgroup.GroupConfig{
+		EnableWithSoftness: true,
+		ShutdownOnNonError: true,
+	})
+
+	grp.Go("server", func(ctx context.Context) error {
+		stdout, _ := telepresenceContext(ctx, "intercept", "--namespace", cs.ns(), svc,
 			"--docker-run", "--port", "8000", "--", "--rm", "-v", abs+":/usr/src/app", tag)
-		stdoutCh <- stdout
-	}()
+		cs.Contains(stdout, "Using Deployment "+svc)
+		return nil
+	})
 
-	expectedOutput := "Hello from intercepted echo-server!"
-	cs.Eventually(
-		// condition
-		func() bool {
-			ctx, cancel := context.WithTimeout(ctx, 500*time.Millisecond)
-			defer cancel()
-			out, err := output(ctx, "curl", "--silent", svc)
-			if err != nil {
-				dlog.Error(ctx, err)
-				return false
-			}
-			dlog.Info(ctx, out)
-			return strings.Contains(out, expectedOutput)
-		},
-		30*time.Second, // waitFor
-		1*time.Second,  // polling interval
-		`body of %q equals %q`, "http://"+svc, expectedOutput,
-	)
-	cancel()
-	cs.Contains(<-stdoutCh, "Using Deployment "+svc)
+	grp.Go("client", func(ctx context.Context) error {
+		expectedOutput := "Hello from intercepted echo-server!"
+		cs.Eventually(
+			// condition
+			func() bool {
+				ctx, cancel := context.WithTimeout(ctx, 500*time.Millisecond)
+				defer cancel()
+				out, err := output(ctx, "curl", "--silent", svc)
+				if err != nil {
+					dlog.Error(ctx, err)
+					return false
+				}
+				dlog.Info(ctx, out)
+				return strings.Contains(out, expectedOutput)
+			},
+			30*time.Second, // waitFor
+			1*time.Second,  // polling interval
+			`body of %q equals %q`, "http://"+svc, expectedOutput,
+		)
+		return nil
+	})
+
+	cs.NoError(grp.Wait())
 }
 
 func (cs *connectedSuite) TestL_LegacySwapDeploymentDoesIntercept() {
@@ -991,27 +994,19 @@ func telepresence(t testing.TB, args ...string) (string, string) {
 
 // telepresence executes the CLI command in-process
 func telepresenceContext(ctx context.Context, args ...string) (string, string) {
-	dlog.Infof(ctx, "running command: %q", append([]string{"telepresence"}, args...))
-
-	cmd := cli.Command(ctx)
+	var stdout, stderr strings.Builder
 
-	stdout := new(strings.Builder)
-	cmd.SetOut(io.MultiWriter(
-		stdout,
-		dlog.StdLogger(dlog.WithField(ctx, "stream", "stdout"), dlog.LogLevelInfo).Writer(),
-	))
+	configDir, _ := filelocation.AppUserConfigDir(ctx)
+	logDir, _ := filelocation.AppUserLogDir(ctx)
 
-	stderr := new(strings.Builder)
-	cmd.SetErr(io.MultiWriter(
-		stderr,
-		dlog.StdLogger(dlog.WithField(ctx, "stream", "stderr"), dlog.LogLevelInfo).Writer(),
-	))
+	cmd := dexec.CommandContext(ctx, client.GetExe(), args...)
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	cmd.Env = append(os.Environ(),
+		"DEV_TELEPRESENCE_CONFIG_DIR="+configDir,
+		"DEV_TELEPRESENCE_LOG_DIR="+logDir)
 
-	cmd.SetArgs(args)
-	if err := cmd.ExecuteContext(ctx); err != nil {
-		fmt.Fprintln(cmd.ErrOrStderr(), err)
-	}
+	_ = cmd.Run()
 
-	dlog.Infof(ctx, "command terminated %q", append([]string{"telepresence"}, args...))
 	return strings.TrimSpace(stdout.String()), strings.TrimSpace(stderr.String())
 }
diff --git a/pkg/client/cli/util.go b/pkg/client/cli/util.go
index 49593475..dbb8b65f 100644
--- a/pkg/client/cli/util.go
+++ b/pkg/client/cli/util.go
@@ -49,13 +49,24 @@ func kubeFlagMap() map[string]string {
 //
 //  - Makes the connector.Connect gRPC call to set up networking
 func withConnector(cmd *cobra.Command, retain bool, f func(context.Context, connector.ConnectorClient, *connector.ConnectInfo) error) error {
-	return cliutil.WithDaemon(cmd.Context(), dnsIP, func(ctx context.Context, daemonClient daemon.DaemonClient) error {
-		if cliutil.DidLaunchDaemon(ctx) && !retain {
-			defer func() { _ = cliutil.QuitDaemon(dcontext.WithoutCancel(ctx)) }()
+	return cliutil.WithDaemon(cmd.Context(), dnsIP, func(ctx context.Context, daemonClient daemon.DaemonClient) (err error) {
+		if cliutil.DidLaunchDaemon(ctx) {
+			defer func() {
+				if err != nil || !retain {
+					_ = cliutil.QuitDaemon(dcontext.WithoutCancel(ctx))
+				}
+			}()
 		}
-		return cliutil.WithConnector(ctx, func(ctx context.Context, connectorClient connector.ConnectorClient) error {
-			if cliutil.DidLaunchConnector(ctx) && !retain {
-				defer func() { _ = cliutil.QuitConnector(dcontext.WithoutCancel(ctx)) }()
+		return cliutil.WithConnector(ctx, func(ctx context.Context, connectorClient connector.ConnectorClient) (err error) {
+			if cliutil.DidLaunchConnector(ctx) && !cliutil.DidLaunchDaemon(ctx) {
+				// Don't shut down the connector if we're shutting down the daemon.
+				// The daemon will shut down the connector for us, and if we shut it
+				// down early the daemon will get upset.
+				defer func() {
+					if err != nil || !retain {
+						_ = cliutil.QuitConnector(dcontext.WithoutCancel(ctx))
+					}
+				}()
 			}
 			connInfo, err := setConnectInfo(ctx, cmd.OutOrStdout())
 			if err != nil {
diff --git a/pkg/client/logging/formatter.go b/pkg/client/logging/formatter.go
index 864f0f10..1c317a9d 100644
--- a/pkg/client/logging/formatter.go
+++ b/pkg/client/logging/formatter.go
@@ -47,12 +47,24 @@ func (f *Formatter) Format(entry *logrus.Entry) ([]byte, error) {
 		for key := range data {
 			keys = append(keys, key)
 		}
-		sort.Strings(keys)
+		sort.Slice(keys, func(i, j int) bool {
+			orders := map[string]int{
+				"dexec.pid":    -4,
+				"dexec.stream": -3,
+				"dexec.data":   -2,
+				"dexec.err":    -1,
+			}
+			iOrd := orders[keys[i]]
+			jOrd := orders[keys[j]]
+			if iOrd != jOrd {
+				return iOrd < jOrd
+			}
+			return keys[i] < keys[j]
+		})
 		for _, key := range keys {
-			val := data[key]
-			fmt.Fprintf(b, " %s=%+v", key, val)
+			val := fmt.Sprintf("%+v", data[key])
+			fmt.Fprintf(b, " %s=%q", key, val)
 		}
-		b.WriteByte(')')
 	}
 
 	if entry.HasCaller() && strings.HasPrefix(entry.Caller.File, thisModule+"/") {
```

thallgren commented on pkg/client/logging/formatter.go:
```diff
@@ -47,12 +47,24 @@ func (f *Formatter) Format(entry *logrus.Entry) ([]byte, error) {
 		}
-		sort.Strings(keys)
+		sort.Slice(keys, func(i, j int) bool {
+			orders := map[string]int{
```
You really trust the compiler to optimize this? I see how it could do that, given that it's like a tightly scoped const, but is that a general assumption that can be made?

LukeShu replied:
I trust that even if the compiler doesn't optimize it that it will be fast enough.

thallgren replied:
Yeah, it's hardly on a critical path :-). Just looks a bit weird to recreate a map inside a comparator.

thallgren reviewed (approved):
Looks great. Hit it!

thallgren commented on pkg/client/cli/cliutil/daemon.go:
```diff
@@ -0,0 +1,186 @@
+	// logging, using dexec would just be extra overhead.
+	"os/exec"
+
+	"github.com/kballard/go-shellquote"
```
We don't really need this extra dep. We have `logging.ShellString(exe, args)` instead of `shellquote.Join(args...)`. I'm changing that back in my Windows PR so you don't need to change it here.
The quoting is different on windows and I've adjusted the `logging.ShellString()` to cope with that.

thallgren commented on pkg/client/cli/cliutil/daemon.go:
```diff
@@ -0,0 +1,186 @@
+		// re-displayed if they typo their password, and (2) it might be ignored anyway
+		// depending on `passprompt_override` in `/etc/sudoers`.  So we'll do a pre-flight
+		// `sudo --non-interactive true` to decide whether to display it.
+		needPwCmd := dexec.CommandContext(ctx, "sudo", "--non-interactive", "--true")
```
```suggestion
		needPwCmd := dexec.CommandContext(ctx, "sudo", "--non-interactive", "true")
```

LukeShu replied:
Oops, I had meant to change that from `true` to `--validate`, but I only made half the change!

thallgren reviewed (changes requested):
I found a minor glitch.

LukeShu committed:
.golangci.yml: Add missing ".com" in some messages

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
Use `pkg/client/logging.ShellString` instead of `github.com/kballard/go-shellquote`

Thomas tells me that go-shellquote does the wrong thing on Windows, but
that he has made ShellString do the right thing.

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu commented:
v3:
```patch
diff --git a/.golangci.yml b/.golangci.yml
index d54b1acb..5e0212cb 100644
--- a/.golangci.yml
+++ b/.golangci.yml
@@ -31,8 +31,9 @@ linters-settings:
     packages-with-error-message:
       - os/exec:                    "Use `github.com/datawire/dlib/dexec` instead of `os/exec`"
       - github.com/golang/protobuf: "Use `google.golang.org/protobuf` instead of `github.com/golang/protobuf`"
-      - github/datawire/dlib/dutil: "Use either `github/datawire/dlib/derror` or `github/datawire/dlib/dhttp` instead of `github/datawire/dlib/dutil`"
-      - golang.org/x/net/http2/h2c: "Use `github/datawire/dlib/dhttp` instead of `golang.org/x/net/http2/h2c`"
+      - github/datawire/dlib/dutil: "Use either `github.com/datawire/dlib/derror` or `github.com/datawire/dlib/dhttp` instead of `github.com/datawire/dlib/dutil`"
+      - golang.org/x/net/http2/h2c: "Use `github.com/datawire/dlib/dhttp` instead of `golang.org/x/net/http2/h2c`"
+      - github.com/kballard/go-shellquote: "Use `github.com/telepresenceio/telepresence/pkg/client/logging.ShellString` instead of `github.com/kballard/go-shellquote.Join`"
 
   gocyclo:
     min-complexity: 35
diff --git a/go.mod b/go.mod
index b208ec99..6ccd1aff 100644
--- a/go.mod
+++ b/go.mod
@@ -10,7 +10,6 @@ require (
 	github.com/godbus/dbus/v5 v5.0.4-0.20201218172701-b3768b321399
 	github.com/google/go-cmp v0.5.5
 	github.com/google/uuid v1.1.2
-	github.com/kballard/go-shellquote v0.0.0-20180428030007-95032a82bc51
 	github.com/miekg/dns v1.1.35
 	github.com/pkg/browser v0.0.0-20180916011732-0a3d74bf9ce4
 	github.com/pkg/errors v0.9.1
diff --git a/pkg/client/cli/cliutil/connector.go b/pkg/client/cli/cliutil/connector.go
index 117ce193..e4626567 100644
--- a/pkg/client/cli/cliutil/connector.go
+++ b/pkg/client/cli/cliutil/connector.go
@@ -15,7 +15,6 @@ import (
 	// logging, using dexec would just be extra overhead.
 	"os/exec"
 
-	"github.com/kballard/go-shellquote"
 	"google.golang.org/grpc"
 	grpcCodes "google.golang.org/grpc/codes"
 	grpcStatus "google.golang.org/grpc/status"
@@ -24,6 +23,7 @@ import (
 	"github.com/datawire/dlib/dgroup"
 	"github.com/telepresenceio/telepresence/rpc/v2/connector"
 	"github.com/telepresenceio/telepresence/v2/pkg/client"
+	"github.com/telepresenceio/telepresence/v2/pkg/client/logging"
 	"github.com/telepresenceio/telepresence/v2/pkg/filelocation"
 )
 
@@ -40,10 +40,10 @@ func launchConnector() error {
 	}
 
 	if err := cmd.Start(); err != nil {
-		return fmt.Errorf("%s: %w", shellquote.Join(args...), err)
+		return fmt.Errorf("%s: %w", logging.ShellString(args[0], args[1:]), err)
 	}
 	if err := cmd.Process.Release(); err != nil {
-		return fmt.Errorf("%s: %w", shellquote.Join(args...), err)
+		return fmt.Errorf("%s: %w", logging.ShellString(args[0], args[1:]), err)
 	}
 
 	return nil
diff --git a/pkg/client/cli/cliutil/daemon.go b/pkg/client/cli/cliutil/daemon.go
index c130a005..b0ae8902 100644
--- a/pkg/client/cli/cliutil/daemon.go
+++ b/pkg/client/cli/cliutil/daemon.go
@@ -13,13 +13,13 @@ import (
 	// logging, using dexec would just be extra overhead.
 	"os/exec"
 
-	"github.com/kballard/go-shellquote"
 	"google.golang.org/grpc"
 	empty "google.golang.org/protobuf/types/known/emptypb"
 
 	"github.com/datawire/dlib/dexec"
 	"github.com/telepresenceio/telepresence/rpc/v2/daemon"
 	"github.com/telepresenceio/telepresence/v2/pkg/client"
+	"github.com/telepresenceio/telepresence/v2/pkg/client/logging"
 	"github.com/telepresenceio/telepresence/v2/pkg/filelocation"
 )
 
@@ -61,16 +61,16 @@ func launchDaemon(ctx context.Context, dnsIP string) error {
 		// use `sudo`'s `--prompt` flag for this because (1) we don't want it to be
 		// re-displayed if they typo their password, and (2) it might be ignored anyway
 		// depending on `passprompt_override` in `/etc/sudoers`.  So we'll do a pre-flight
-		// `sudo --non-interactive true` to decide whether to display it.
-		needPwCmd := dexec.CommandContext(ctx, "sudo", "--non-interactive", "--true")
+		// `sudo --non-interactive --validate` to decide whether to display it.
+		needPwCmd := dexec.CommandContext(ctx, "sudo", "--non-interactive", "--validate")
 		needPwCmd.DisableLogging = true
 		if err := needPwCmd.Run(); err != nil {
-			fmt.Printf("Need root privileges to run: %s\n", shellquote.Join(args...))
+			fmt.Printf("Need root privileges to run: %s\n", logging.ShellString(args[0], args[1:]))
 			// `sudo` won't be able to read the password from the terminal when we run
-			// it with Setpgid=true, so do a pre-flight `sudo true` to read the
+			// it with Setpgid=true, so do a pre-flight `sudo --validate` to read the
 			// password, and then enforce that being re-used by passing
 			// `--non-interactive`.
-			pwCmd := dexec.CommandContext(ctx, "sudo", "true")
+			pwCmd := dexec.CommandContext(ctx, "sudo", "--validate")
 			pwCmd.DisableLogging = true
 			if err := pwCmd.Run(); err != nil {
 				return err
@@ -87,10 +87,10 @@ func launchDaemon(ctx context.Context, dnsIP string) error {
 	}
 
 	if err := cmd.Start(); err != nil {
-		return fmt.Errorf("%s: %w", shellquote.Join(args...), err)
+		return fmt.Errorf("%s: %w", logging.ShellString(args[0], args[1:]), err)
 	}
 	if err := cmd.Process.Release(); err != nil {
-		return fmt.Errorf("%s: %w", shellquote.Join(args...), err)
+		return fmt.Errorf("%s: %w", logging.ShellString(args[0], args[1:]), err)
 	}
 
 	return nil
```
